Article about Behavior-Driven Development with SpecFlow and .NET (#40)
zuhlke/engineering-posts: drafts/bdd-specflow.md

    public void WhenTheUserEntersInvalidLoginCredentials()
    {
        _app.LoginService.EnterCredentials("nope", "nope");
    }

    [When(@"the user clicks on the login button")]
    public void WhenTheUserClicksOnTheLoginButton()
    {
        try
        {
            _app.LoginService.Login();
            _app.NavigationService.GoToInternalPage();
        }
        catch (Exception ex)
        {
            _exception = ex;
        }
    }

    [Then(@"an error message is shown")]
    public void ThenAnErrorMessageIsShown()
    {
        Assert.IsNotNull(_exception);
    }

    [Then(@"the user stays on the login page")]
    public void ThenTheUserStaysOnTheLoginPage()
    {
        Assert.IsTrue(_app.NavigationService.IsOnLoginPage);
    }

    [When(@"the user enters valid login credentials")]
    public void WhenTheUserEntersValidLoginCredentials()
    {
        _app.LoginService.EnterCredentials("admin", "pass");
    }

    [Then(@"the user is redirected to an internal page")]
    public void ThenTheUserIsRedirectedToAnInternalPage()
    {
        Assert.IsTrue(_app.NavigationService.IsOnInternalPage);
    }
}
```

## Executing the tests

> Build your project and then the test runner should find the new tests and be able to execute them. For every feature file, there will an auto-generated code after the build.

Reviewer: It seems like there is a word missing after the comma: `there will an auto-generated code`. 
```suggestion
Build your project and then the test runner should find the new tests and be able to execute them. For every feature file, there will be an auto-generated code after the build.
```

Author: I even added two words :)

Reviewer: Perfect 😅👍

---
title: How to use Behaviour-Driven Development (BDD) with C# and SpecFlow
subtitle: Here is an introductory guide on how to use Behaviour-Driven Development (BDD) with C# and SpecFlow in a .NET application.
domain: software-engineering-corner.hashnode.dev
tags: web-development, azure, build-tool, sass
cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1682689039164/_2bkGoHwv.webp?auto=format

> publishAs: alexander-duschek

Author: Thanks. I just noticed, I also forgot to update the tags :)

hideFromHashnodeCommunity: false
ignorePost: true
---

Behaviour-Driven Development (BDD) is a technique of agile software development to ensure that the requirements of the software to be developed are fulfilled. To achieve this a special syntax called Gherkin (see below) is used to write down the specification which then can be translated by tools into test case skeletons. The developers add the test code and write/enhance the application until the test passes.
A big advantage is that stakeholders, customers, business analysts, requirement engineers, all parties of a project can provide tests.

## BDD Key features

- Everyone can provide test cases
- Test cases are easy to read
- Good tool support for automation
- Developers need less domain knowledge
- Developers can focus on writing tests and features instead of finding test cases
- Project documentation on the fly

## What about Test-Driven Development (TDD)?

> Simply put, if a software developer uses the TDD technique, he first creates unit tests based on the requirements or the specification of the software or feature. Afterward, he enhances the software until the tests will eventually pass. For more details read the following article.

Reviewer 2: ```suggestion
Simply put, if a software developer uses the TDD technique, they first create unit tests based on the requirements or the specification of the software or feature. Afterward, they enhance the software until the tests will eventually pass. For more details read the following article.
```

Prefer the use of neutral pronouns.

> https://medium.com/geekculture/test-driven-development-that-every-developer-should-know-about-with-example-c5f68af547a6

Reviewer 2: I would put this link directly into the text like `[This article on Medium](https...) explains TDD in more detail`.

> As you probably noticed, BDD and TDD can be combined. The main difference is that with TDD the developer creates the test cases and with BDD every person related to the project can supply tests with the common Gherkin syntax (see next paragraph).

Reviewer 2: ```suggestion
As you probably noticed, BDD and TDD can be combined. The main difference is that with TDD the developer creates the test cases and with BDD every person related to the project can supply tests with the common Gherkin syntax.
```
I think we can trust in readers noticing the next line ;)

## Gherkin syntax

> The Gherkin syntax is a way to describe can be seen in the image below. It consists of keywords at the beginning of each line followed by explaining text.

Reviewer 2: The first sentence doesn't quite make sense. Looks like some words are left out or at least jumbled. Also, the referred image is missing.

Author: I published this article on Medium and my own blog, it was the top result in Google for quite some time and NOBODY mentioned this error :D

```
Feature: Login
    Ensures that access to internal areas is only provided for authenticated users
    
Scenario: Invalid login credentials prevent access to internal areas
    Given the user has opened the login page
    When the user enters invalid login credentials
    When the user clicks on the login button
    Then an error message is shown
    Then the user stays on the login page
    
Scenario: Valid login credentials allow access to internal areas
    Given the user has opened the login page
    When the user enters valid login credentials
    When the user clicks on the login button
    Then the user is redirected to an internal page
```

Feature describes a high-level software feature and is used to group Scenarios. A Scenario is a test case for the feature. It consists of preconditions (Given), actions (When), and assertions (Then). There are many possibilities to provide more detailed information but the basic set of a feature file looks like this in Gherkin syntax.

> To make it more natural, identical keywords can be chained by and. So instead of

Reviewer 2: I'd highlight "and" somehow, with double quote, italics or inline code. Otherwise it looks like an unfinished sentence at first :D

```
Given ...
Given ...
Given ...
```

you can write

```
Given ...
And ...
And ...
```

Feels like talking, doesn’t it? 😊

## From Text to Code

Ok, we have our text file in Gherkin syntax. To create test case skeletons we need an extension for Visual Studio called [SpecFlow](https://specflow.org/). It’s an open-source project to enable BDD in .NET environments and will save you a lot of typing. The extension is available for [Visual Studio](https://marketplace.visualstudio.com/search?term=specflow&target=VS&category=All%20categories&vsVersion=&sortBy=Relevance) and [Rider](https://plugins.jetbrains.com/plugin/15957-specflow-for-rider). The following guide uses the Visual Studio extension.

You can install the extension from the [Visual Studio Marketplace](https://marketplace.visualstudio.com/search?term=specflow&target=VS&category=All%20categories&vsVersion=&sortBy=Relevance) (or follow this [video guide](https://docs.specflow.org/projects/getting-started/en/latest/GettingStarted/Step1.html) for further assistance).

> SpecFlow works with a .feature file (a simple text file with the ending .feature) that contains the Gherkin text and a binding file (a C# code file) that contains the actual C# test code. Every line in the feature file is mapped to a call of a method in the C# file. Based on the image above our C# file should have 7 methods (some lines are identical and share the same code).

Reviewer 2: "Based on the image _below_"? Or is the image missing?
